torch/__init__.py

```diff
@@ -245,6 +245,15 @@ def typename(o):
     else:
         class_name = o.__class__.__name__
 
+    # Sometimes we want to get the typename of type aliases from the Typing
+    # package. These type aliases don't have a descriptive classname (e.g.
+    # typing._GenericAlias, and also that's not stable across versions) so
+    # we use the repr of the type alias instead.
+    #
+    # TODO: There's gotta be a better way to do this
+    if module == 'typing.':
+        return str(o)
+
```

Comment:
@dzhulgakov @zdevito this seems sketchy to me since it's too permissive in what things in `typing` it matches on, but I didn't find a better way to check "just type aliases". On my python version the type aliases subclass `typing._GenericAlias` but that doesn't seem to be consistent. Any ideas if there's a better way.

Reply:
I think the problem is that `typename(o)` returns the type of the given object. Here you want something that gives a str representation of something that is already a type. Since types are objects too it gets confusing as usages within pytorch codebase are mixed, I see both `typename(obj)` and `typename(type(obj))`.

You probably should create a separate function (sorry for confusion) but you can follow the logic similar to here to handle builtins and stuff.

Reply:
> I think the problem is that typename(o) returns the type of the given object. 

This is only true for the codepath on line 246, all the other codepaths assume that `o` is a type itself. The contract for this function seems to be "given an object, return the name of its type. given a type, return the qualified name of that type"

Reply:
Yeah, it's messed up :) There's https://github.com/python/cpython/blob/master/Lib/typing.py#L156 also but it seems similar to what goes on here.

Reply:
k i'm gonna copy and paste that so i have plausible deniability if something is busted